packages/flutter/lib/src/cupertino/scaffold.dart

```diff
@@ -0,0 +1,157 @@
+  /// provide [CupertinoTabBar]'s tap callbacks to change the active tab.
+  ///
+  /// Tabs' contents is built with the provided [rootPageBuilder] at the active
+  /// tab index.
+  const CupertinoScaffold.tabbed({
+    Key key,
+    this.navigationBar,
+    @required this.tabBar,
+    @required this.rootPageBuilder,
+  }) : assert(tabBar != null),
+       assert(rootPageBuilder != null),
+       child = null,
+       super(key: key);
+
+  /// The [navigationBar], typically a [CupertinoNavigationBar] is drawn at the
+  /// top of the screen.
+  ///
+  /// If translucent, the main content may slide behind it.
+  /// Otherwise, the main content's top margin will be offset by its height.
+  // TODO(xster): document its page transition animation when ready
+  final PreferredSizeWidget navigationBar;
+
+  /// The [tabBar] is a [CupertinoTabBar] drawn at the bottom of the screen
+  /// that lets the user switch between different tabs in the main content area
+  /// when present.
+  ///
+  /// When provided, [CupertinoTabBar.currentIndex] will be ignored and and will
+  /// be managed by the [CupertinoScaffold] to show the currently selected page
+  /// as the active item index. [CupertinoTabBar.onTap] will still be called but
+  /// chained after an automatic handler from [CupertinoScaffold] to change
+  /// the current tab in the main content area.
+  ///
+  /// If translucent, the main content may slide behind it.
+  /// Otherwise, the main content's bottom margin will be offset by its height.
+  final CupertinoTabBar tabBar;
+
+  /// An [IndexedWidgetBuilder] that's called when tabs become active. Used
+  /// when a tabbed scaffold is constructed via the [.tabbed] constructor.
+  ///
+  /// Content can slide under the [navigationBar] or the [tabBar] when they're
+  /// translucent.
+  final IndexedWidgetBuilder rootPageBuilder;
+
+  /// Widget to show in the main content area when the scaffold is used without
+  /// tabs.
+  ///
+  /// Content can slide under the [navigationBar] or the [tabBar] when they're
+  /// translucent.
+  final Widget child;
+
+  @override
+  _CupertinoScaffoldState createState() => new _CupertinoScaffoldState();
+}
+
+class _CupertinoScaffoldState extends State<CupertinoScaffold> {
+  int _currentPage = 0;
+
+  @override
+  Widget build(BuildContext context) {
+    final List<Widget> stacked = <Widget>[];
+
+    // The main content being at the bottom is added to the stack first.
+    if (widget.child != null) {
+      stacked.add(_padMiddle(widget.child));
+    } else if (widget.rootPageBuilder != null) {
+      stacked.add(_padMiddle(
+        new IndexedStack(
+          index: _currentPage,
+          children: new List<Widget>.generate(
+            widget.tabBar.items.length,
+            (int index) {
+              return new Builder(
+                builder: (BuildContext context) {
+                  return widget.rootPageBuilder(context, index);
```

Comment:
Yeah we definitely don't want to use an IndexedStack here.
...
...
(long in-person discussion)
...
...
We concluded we want to use a Stack, lazily inflate the children so that you only inflate them once they've been picked, then keep them in the tree, and you want to use TickerMode and Offstage to keep them cheaper.